hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/SparkMetadataTableRecordIndex.java

```diff
@@ -0,0 +1,227 @@
+      return true;
+    }
+  }
+
+  @Override
+  public boolean isGlobal() {
+    return true;
+  }
+
+  @Override
+  public boolean canIndexLogFiles() {
+    return true;
+  }
+
+  @Override
+  public boolean isImplicitWithStorage() {
+    return false;
+  }
+
+  private <R> HoodieData<HoodieRecord<R>> tagLocationBackToRecords(
+      HoodiePairData<String, HoodieRecordGlobalLocation> keyFilenamePair,
+      HoodieData<HoodieRecord<R>> records) {
+    HoodiePairData<String, HoodieRecord<R>> keyRecordPairs =
+        records.mapToPair(record -> new ImmutablePair<>(record.getRecordKey(), record));
+    // Here as the records might have more data than keyFilenamePairs (some row keys' not found in record index),
+    // we will do left outer join.
+    return keyRecordPairs.leftOuterJoin(keyFilenamePair).values()
+        .map(v -> {
+          HoodieRecord<R> record = v.getLeft();
+          Option<HoodieRecordGlobalLocation> location = Option.ofNullable(v.getRight().orElse(null));
+          if (!location.isPresent()) {
+            // No location found.
+            return record;
+          }
+          // Ensure the partitionPath is also set correctly in the key
+          if (!record.getPartitionPath().equals(location.get().getPartitionPath())) {
+            record = new HoodieAvroRecord(new HoodieKey(record.getRecordKey(), location.get().getPartitionPath()), (HoodieRecordPayload) record.getData());
+          }
+
+          // Perform the tagging. Not using HoodieIndexUtils.getTaggedRecord to prevent an additional copy which is not necessary for this index.
+          record.unseal();
+          record.setCurrentLocation(location.get());
+          record.seal();
+          return record;
+        });
+  }
+
+  /**
+   * Function that lookups a list of keys in a single shard of the record index
+   */
+  private static class RecordIndexFileGroupLookupFunction implements PairFlatMapFunction<Iterator<String>, String, HoodieRecordGlobalLocation> {
+    private final HoodieTable hoodieTable;
+
+    public RecordIndexFileGroupLookupFunction(HoodieTable hoodieTable) {
+      this.hoodieTable = hoodieTable;
+    }
+
+    @Override
+    public Iterator<Tuple2<String, HoodieRecordGlobalLocation>> call(Iterator<String> recordKeyIterator) {
+      List<String> keysToLookup = new ArrayList<>();
+      recordKeyIterator.forEachRemaining(keysToLookup::add);
+
+      // recordIndexInfo object only contains records that are present in record_index.
+      Map<String, HoodieRecordGlobalLocation> recordIndexInfo = hoodieTable.getMetadataTable().readRecordIndex(keysToLookup);
+
+      HoodieTableMetaClient metaClient = hoodieTable.getMetaClient();
+      HoodieTimeline commitsTimeline = metaClient.getCommitsTimeline().filterCompletedInstants();
+      return recordIndexInfo.entrySet().stream()
+          .filter(e -> HoodieIndexUtils.checkIfValidCommit(commitsTimeline, e.getValue().getInstantTime()))
```

Comment:
I feel this might be adding some overhead to the latency. Is there any optimization we can do here?

Reply:
synced up w/ Surya. so far, could not come up w/ valid scenario where this is strictly required.

Reply:
Yes, not needed. removed.